ra/registration-authority.go

```diff
@@ -750,17 +750,36 @@ func (ra *RegistrationAuthorityImpl) UpdateAuthorization(base core.Authorization
 		return
 	}
 
-	// Copy information over that the client is allowed to supply
 	authz = base
 	if challengeIndex >= len(authz.Challenges) {
 		err = core.MalformedRequestError(fmt.Sprintf("Invalid challenge index: %d", challengeIndex))
 		return
 	}
 
-	authz.Challenges[challengeIndex].KeyAuthorization = response.KeyAuthorization
+	ch := &authz.Challenges[challengeIndex]
+
+	// Copy information over that the client is allowed to supply
+	ch.ProvidedKeyAuthorization = response.ProvidedKeyAuthorization
+
+	if response.Type != "" && ch.Type != response.Type {
```

Comment:
This is a newly introduced check, right? I like it in practice but we want to make sure we're not going to spontaneously break any currently-working clients.